core/types/celo_receipt_test.go

```diff
@@ -101,3 +102,119 @@ func checkStorageRLPEncodeDecodeConsistency(r *ReceiptForStorage, t *testing.T)
 
 	require.EqualValues(t, r, &r2)
 }
+
+// Tests that the effective gas price is correctly derived for different transaction types, in different scenarios.
+func TestReceiptEffectiveGasPriceDerivation(t *testing.T) {
+	gasPrice := big.NewInt(1000)
+	gasFeeCap := big.NewInt(800)
+	gasTipCap := big.NewInt(100)
+	config := params.TestChainConfig
+	config.GingerbreadBlock = big.NewInt(1)
+	// Receipt base fee is the base fee encoded in the receipt which will be set post cel2 for CeloDynamicFeeTxV2 types.
+	receiptBaseFee := big.NewInt(50)
+
+	t.Run("LegacyTx", func(t *testing.T) {
+		testNonDynamic(t, NewTransaction(0, common.Address{}, big.NewInt(0), 0, gasPrice, nil), gasPrice)
+	})
+	t.Run("AccessListTx", func(t *testing.T) {
+		testNonDynamic(t, NewTx(&AccessListTx{GasPrice: gasPrice}), gasPrice)
+	})
+	t.Run("DynamicFeeTx", func(t *testing.T) {
+		tx := NewTx(&DynamicFeeTx{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap})
+		testDynamic(t, tx, nil)
+	})
+	t.Run("CeloDynamicFeeTx", func(t *testing.T) {
+		tx := NewTx(&CeloDynamicFeeTx{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap})
+		testDynamic(t, tx, nil)
+		tx = NewTx(&CeloDynamicFeeTx{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap, FeeCurrency: &common.Address{}})
+		testDynamicWithFeeCurrency(t, tx, nil)
+	})
+	t.Run("CeloDynamicFeeTxV2", func(t *testing.T) {
+		tx := NewTx(&CeloDynamicFeeTxV2{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap})
+		testDynamic(t, tx, nil)
+		testDynamic(t, tx, receiptBaseFee)
+		tx = NewTx(&CeloDynamicFeeTxV2{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap, FeeCurrency: &common.Address{}})
+		testDynamicWithFeeCurrency(t, tx, nil)
+		testDynamicWithFeeCurrency(t, tx, receiptBaseFee)
+	})
+	t.Run("CeloDenominatedTx", func(t *testing.T) {
+		tx := NewTx(&CeloDenominatedTx{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap})
+		testDynamic(t, tx, nil)
+		tx = NewTx(&CeloDynamicFeeTxV2{GasFeeCap: gasFeeCap, GasTipCap: gasTipCap, FeeCurrency: &common.Address{}})
+		testDynamicWithFeeCurrency(t, tx, nil)
+	})
+}
+
+func testNonDynamic(t *testing.T, tx *Transaction, receiptBaseFee *big.Int) {
+	// Non dynamic txs should always have the gas price defined in the tx.
+	config := params.TestChainConfig
+	config.GingerbreadBlock = big.NewInt(1)
+	preGingerbreadBlock := uint64(0)
+	postGingerbreadBlock := uint64(2)
```

Comment:
You could make those global variables instead of redefining in every function

Reply:
The thing is that I can't set `config.GingerbreadBlock = big.NewInt(1)` in a var block, so I'd be reduced to only setting the config and the pre and post gingerbread blocks but then having to modify the config elsewhere.

Modifying a global variable is a bad idea in case other test rely on the unmodified versions. So given that only the pre and post gingerbread blocks should be made global, but then since the pre and post gingerbread blocks really only have meaning in the context of the modified config, it didn't make sense to me to make them global either.